amethyst_rendy/src/bundle.rs

```diff
@@ -0,0 +1,950 @@
+                    .get_pass_node_raw(image.target())
+                    .expect("Image without target node");
+                self.add_dep(node);
+                i
+            })
+        })
+    }
+
+    /// Add explicit dependency on another node.
+    ///
+    /// This is done automatically when you use `get_image`.
+    pub fn add_dep(&mut self, node: NodeId) {
+        if !self.deps.contains(&node) {
+            self.deps.push(node);
+        }
+    }
+
+    /// Access underlying rendy's GraphBuilder directly.
+    /// This is useful for adding custom rendering nodes
+    /// that are not just standard graphics render passes,
+    /// e.g. for compute dispatch.
+    pub fn graph(&mut self) -> &mut GraphBuilder<B, Resources> {
+        self.plan_context.graph()
+    }
+
+    /// Retrieve render target metadata, e.g. size.
+    pub fn target_metadata(&self, target: Target) -> Option<TargetMetadata> {
+        self.plan_context.target_metadata(target)
+    }
+
+    /// Access computed NodeId of render target.
+    pub fn get_node(&mut self, target: Target) -> Result<NodeId, Error> {
+        self.plan_context.get_node(target)
+    }
+}
+
+/// An identifier for output image of specific render target.
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
+pub enum TargetImage {
+    /// Select target color output with given index.
+    Color(Target, usize),
+    /// Select target depth output.
+    Depth(Target),
+}
+
+impl TargetImage {
+    /// Retrieve target identifier for this image
+    pub fn target(&self) -> Target {
+        match self {
+            TargetImage::Color(target, _) => *target,
+            TargetImage::Depth(target) => *target,
+        }
+    }
+}
+
+/// Set of options required to create an image node in render graph.
+#[derive(Debug, Clone, PartialEq)]
+pub struct ImageOptions {
+    /// Image kind and size
+    pub kind: hal::image::Kind,
+    /// Number of mipmap levels
+    pub levels: hal::image::Level,
+    /// Image format
+    pub format: hal::format::Format,
+    /// Clear operation performed once per frame.
+    pub clear: Option<hal::command::ClearValue>,
+}
+
+/// Definition of render target color output image.
+#[derive(Debug)]
+pub enum OutputColor<B: Backend> {
+    /// Render to image with specified options
+    Image(ImageOptions),
+    /// Render directly to a window surface.
+    Surface(Surface<B>, Option<hal::command::ClearValue>),
+}
+
+/// Definition for set of outputs for a given render target.
+#[derive(Debug)]
+pub struct TargetPlanOutputs<B: Backend> {
+    /// List of target color outputs with options
+    pub colors: Vec<OutputColor<B>>,
+    /// Settings for optional depth output
+    pub depth: Option<ImageOptions>,
+}
+
+#[derive(derivative::Derivative)]
+#[derivative(Debug(bound = ""))]
+struct TargetPlan<B: Backend> {
+    key: Target,
+    #[derivative(Debug = "ignore")]
+    extensions: Vec<Box<dyn FnOnce(&mut TargetPlanContext<'_, B>) -> Result<(), Error> + 'static>>,
```

Comment:
If plugin would register extensions with names it will be easier debuggable

Reply:
Extension types might have names instead. Might be added in the future, I don't see exactly a place to use that right now.